x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java

```diff
@@ -152,12 +146,124 @@ public void init() {
             Set.of("foo")
         );
 
-        analyzer = new Analyzer(
-            new AnalyzerContext(config, functionRegistry, getIndexResult, enrichResolution),
+        return new Analyzer(
+            new AnalyzerContext(config, new EsqlFunctionRegistry(), getIndexResult, enrichResolution),
             new Verifier(new Metrics())
         );
     }
 
+    private Analyzer allFieldMappingAnalyzer() {
+        var mapping = loadMapping("mapping-alltypes.json");
+        EsIndex test = new EsIndex("test", mapping);
+        IndexResolution getIndexResult = IndexResolution.valid(test);
+
+        return new Analyzer(new AnalyzerContext(config, new EsqlFunctionRegistry(), getIndexResult, null), new Verifier(new Metrics()));
+    }
+
+    private record ImplicitCastCase(String datatype, String expression) {}
+
+    /**
+     * Expects the filter to be trivialized to a TRUE or FALSE literal, e.g.:
+     * LimitExec[500[INTEGER]]
+     * \_ExchangeExec[[],false]
+     *   \_EsQueryExec[test], query[][_doc{f}#19], limit[500], sort[] estimatedRowSize[4]
+     */
+    public void testTrivialFilterForUnsupportedImplicitCast() {
```

Comment:
The tests are duplicated across the rest tests and here - move the beef of the tests here and add a couple on the rest layer.
Since this impacts Lucene, another approach is to create a simple index, test the generated query and then look at its output. However at that point your retesting the conversion and not your rule.